detailed instructions for release procedure (including testing recipes and other such fun stuff) (#3032)
ESMValGroup/ESMValTool: doc/sphinx/source/community/detailed_release_procedure.rst

.. _detailed-release-procedure:

Release: recipes runs and comparison
====================================

The release procedure for ESMValTool is a fairly involved process (at the moment), so it
is important to be very well organized and to have documented each procedural steps, so that
the next release manager can follow said steps, and finalize the release without any delays.

The workflow below assumes an ESMValCore release candidate, or a completed stable release, have been released
and deployed on conda-forge and PyPI; it also assumes the release manager has access to accounts on DKRZ/Levante.

Below is a list of steps that the release manager, together with the previous release manager, should go through before the actual release;

> these include testing the new code by running all available recipes in yhe `main` branch, and comparing the output against

Reviewer: ```suggestion
these include testing the new code by running all available recipes in the ``main`` branch, and comparing the output against
```

the previous release.

Open an issue on GitHub
-----------------------

> First, open an issue on GitHub where the testing workflow before the release is documented (see example https://github.com/ESMValGroup/ESMValTool/issues/2881).

Reviewer: A discussion might be more suitable for this purpose, as it allows organized conversation about multiple different topics. E.g. for every broken recipe, a new thread could be started.

Reviewer 2: I'm not so convinced about the usage of discussions. If we do several rounds of recipe testing (likely to be the case for v2.8), it might be simpler to open/close issues for each round. It might not be clear for the recipe maintainers which discussion is the active one and threads would spread across several discussions.

Author: I too agree with Remi (and implicitly, with myself haha) - Issues are: 1. more visible and 2. closable and forgettable and searchable, plus - the "reply to a reply" format in Discussions may lead to bits that are missed

Reviewer: It;s up to you, but I personally find an issue becomes very difficult to read if you're trying to discuss several issues at the same time in it. If this happens at a high pace (several messages a day), it becomes quite impossible to follow for relative outsiders such as recipe maintainers or other core team members not as actively involved in the release. If you really prefer issues, my recommendation would be to open an issue for each broken recipe separately and try to limit the conversation in the overview issue somehow.

Name it something relevant like "Recipe testing and comparison for release 2.x.x", and populate the issue description with information
about where the testing is taking place, what tools are used, and what versions, here are some suggestions:

Where the work is done e.g. work done on DKRZ/Levante:

- submit files in `/home/b/b382109/submit` (for v2.7.0)
- output in `/scratch/b/b382109/esmvaltool_output` (for v2.7.0)

We should document various utilities' versions so that the work can be reproduced in case there
is an issue, or release work needs to be picked up mid-release by another release manager:

- documenting `conda`/`mamba` versions:

.. code-block:: bash

  mamba --version

- documenting `git` branch and its state:

.. code-block:: bash

  git status

Furthermore, the runtime environment needs to be documented:

.. code-block:: bash

  mamba env create -n tool2xxTest -f environment.yml

>   conda activate tool2xxTest

Reviewer: This looks like part of the installation instructions. It will be easier to maintain this documentation if it just includes a link instead of some duplicate text.

Author: good point! Let me see how I can wedge that in

Author: I have actually just removed it, the RM should really know how to install by now :laughing:

Make a copy of the environment file, and attach it in the release testing issue; to
record the environment in a yaml file use e.g. `conda env export > ToolEnv270Test.txt`.

Mention any special data movement one needs to perfom to run the tests e.g.:
for v2.7.0 the release manager moved the autoassess-specific files to

> `/home/b/$USER/autoassess_files` on DKRZ/Levante.

Reviewer: As stated in our [broken recipe policy](https://esmvaltool--3032.org.readthedocs.build/en/3032/community/broken_recipe_policy.html), all data should be publicly available and instructions to obtain it should be part of the recipe documentation. Why is this needed?

Author: ha! If only that were true :grin: I'd put this up for discussion - I need to leave this item at least for now so @remi-kazeroni can get his data on Levante (although I know Remi knows where it is, but just in case)

Reviewer 3: IIRC we don't have redistribution rights on these files, so I don't think we should do that. There is a better solution (discussed in #2309) but I haven't managed to implement it yet. Perhaps the best option in the short term is to suppress/remove these broken recipes until we fix them?

Reviewer 2: Thanks for your answer @alistairsellar. I will take a look at those recipes when I run them for the release. If these recipes can't be run because of non-publicly available files, I will follow your advice and mark these as broken.

Modifications to configuration files need to be documented as well: for example,
for v2.7.0 the release manager had to modify `config-user.yml` file by
adding a number of extra paths for DKRZ/Levante-specific data pools:

.. code-block:: yaml

  CMIP6:
    - /work/bd0854/DATA/ESMValTool2/CMIP6_DKRZ
    - /work/bd0854/DATA/ESMValTool2/download/CMIP6
  CMIP5:
    - /work/bd0854/DATA/ESMValTool2/CMIP5_DKRZ

>     - /work/bd0854/b309141/additional_CMIP5

Reviewer: Does this directory contain retracted data that is no longer available on ESGF? It is against our [broken recipe policy](https://esmvaltool--3032.org.readthedocs.build/en/3032/community/broken_recipe_policy.html) to include recipes using such data in a release. Therefore I would recommend against using anything but the main data pool from the cluster (which will be hopefully cleaned up by the cluster admins) and some recently downloaded data from ESGF.

Reviewer 2: At DKRZ, the problem is not so much with CMIP6 data but rather with the CMIP5 pool that is not really maintained and from which lots of data have been removed last year to free disk space. So I'd like to continue using the `/work/bd0854/DATA/ESMValTool2/download/CMIPX`, otherwise results in lots of downloads for every testing round.

Author: indeed - relates to the comment above about publicly available data

Reviewer: I can understand that, though some occasional cleanup might be good. But what about `/work/bd0854/b309141/additional_CMIP5`? It looks like someone's home directory?

    - /work/bd0854/DATA/ESMValTool2/download/cmip5/output1
    - /work/bd0854/DATA/ESMValTool2/download/cmip5

Submit run scripts - test recipe runs
-------------------------------------

We are now ready to start running all the available recipes, to compare output against previous release. Running is currently done

> via batch scripts submitted to a schedulers (SLURM). Generate the submission scripts using the `generate.py` Python script;

Reviewer: ```suggestion
via batch scripts submitted to a schedulers (SLURM). Generate the submission scripts using the ``generate.py`` Python script;
```

> you can find a copy of the script either in `/home/b/b382109/Tool_Release_270_Scripts` or

Reviewer: I would recommend against adding people's home directories to our public documentation. This is likely to become outdated quickly and maybe you don't need everyone and their dog to start having a look at that.

Reviewer 2: I have just created a directory in `work` named `/work/bd0854/esmvaltool_releases` to store these kinds of files. Would it be an option to place the submission scripts and other scripts inside that common directory instead of people's home directories? Feel free to add subdirectories in it with explicit names, e.g. `sbatch_scripts`, `output_parsing_scripts`, ...

Author: yes I agree and thanks, Remi - you mind adding that info here pls?

Reviewer 2: Sure, will do that when I get to these steps during the release. I'll have a better idea of how things could be organized in a more sustainable way

Reviewer: I don't want to sound harsh, but code should really be stored in a version controlled repository, not in some directory on some computer. If you do not feel comfortable having it on the ESMValTool repository, we could create a new repository for testing scripts.

Reviewer 2: @bouweandela, you're fully right. These scripts can be found in the draft PR #2883. Before deciding something about these scripts, we first need to agree on a common approach to run all recipe for the testing rounds. This will be discussed next Tuesday at the Tech Lead Team Meeting, see https://github.com/ESMValGroup/Community/discussions/79#discussioncomment-5086745. 

I personally think that these scripts are our best option for the v2.8 release and the near future. These have proven to be very useful for the past 2 releases. If we all agree on that, I'm happy to take a look at #2883, update it if needed and merge it before the release to have the scripts in v2.8. We would then update the documentation for the release procedure accordingly. Once a more advanced testing framework like the recipe test workflow is published, we could think of retiring these scripts.

Reviewer: See also the comment here: https://github.com/ESMValGroup/ESMValTool/pull/3032#discussion_r1116639801

> in the draft Pull Request https://github.com/ESMValGroup/ESMValTool/pull/2883.

Reviewer: This is something that needs to be discussed with the @ESMValGroup/technical-lead-development-team. Do we start our own set of scripts or can we use an established job submission system?

Author: this uses SLURM (as job submission system/scheduler) - the problem is how we set up the jobs - either with these sort of scripts or with automated test scripts (formerly started as development by the MO, unceremoniously postponed until further notice)

Reviewer: Yes, but there is existing software that can write such scripts (e.g. cylc, [xenon](https://github.com/xenon-middleware/pyxenon), [Dirac](https://dirac.readthedocs.io/en/latest/UserGuide/GettingStarted/UserJobs/DiracAPI/index.html), and probably a lot more). Writing your own scripts is reinventing the wheel and should only be done for very simple things.

> You will have to set the name of your environment, your email address (if you want to get email notifications for successful/failed jobs) and the name of the directory you want to store the job outputs. The name of the account is the same (`bk1088`), and the default partition is set to `compute`.

Reviewer: I wonder if we can still use that account after IS-ENES3 ends

Author: no clue - we need to ask, I can email DKRZ?

Author: I just plopped a note there, more for us to remember to check, then we can change the note to a warning

Reviewer 2: I can take care of that. I think I know who to ask, I'll cc you

Some recipes need other job requirements, you can add their headers in the `SPECIAL_RECIPES` dictionary. Otherwise the header will be written following the template that is written in the lines below. If you want to exclude recipes, you can do so by uncommenting the exclude lines.

The launch scripts will be saved in the same directory you execute the script from. You can find the ones used for the v2.7.0 release in `/home/b/b382109/submit`

To submit the scripts use the `sbatch` submit scripts (that make use of the SLURM scheduler) produced by `generate.py`,
and execute them like any other shell script. You can check the status of your BATCH queue by invoking:

.. code-block:: bash

>   squeue -u b382109

Reviewer: ```suggestion
  squeue -u $USER
```

  As soon as a job is finished, the next one should start

Submit the batch scripts that will run all recipes. Assemble some statistics so that issues with certain recipes
can be followed-up, and document this information in the release issue, examples:

- number of successfully run recipes
- number of failed recipes with Diagnostic error (can they be fixed? Can the fixes be included in the release?)
- number of recipes that are missing data
- number of recipes that have various other issues (and document them)

To parse the output of all these runs use the `parse_recipes_output.py` Python script, included at the
same locations where the generation script is.

Running the comparison
----------------------

To compare the newly produced output from running all recipes, follow these steps below.

Login and access to the DKRZ esmvaltool virtual machine (VM) - results from recipe runs
are stored on the VM; login with:

.. code-block:: bash

  ssh user@esmvaltool.dkrz.de

where `user` is your DKRZ/Levante user name; then get and install miniconda on the VM, and
if you have a Miniconda installer already downloaded in your Levante $HOME

.. code-block:: bash

>   scp Miniconda3-py39_4.12.0-Linux-x86_64.sh user@esmvaltool.dkrz.de:~

Reviewer: Please mention that conda environments should not be created in the home directory because it is on a very small disk, but rather in a directory with your username under `/mnt/esmvaltool_disk2/work/`

Author: very good point indeed! Juast added a warning about that

Next, we need to set up the input files

.. note::

  If you wrote recipe runs output to Levante's `/scratch` partition, be aware that
  the data will be removed after two weeks, so you will have to move the output data
  to the `/work` partition, via e.g. a `nohup` job:

  .. code-block:: bash

>     nohup cp -r /scratch/b/$USER/esmvaltool_output/* /work/bd0854/b382109/v2xx

Author: changed it! Thanks, bud!

  where `bd0854/b382109` is the project location in `work`

The `/work` partition is visible by the VM so you can run the compare tool straight on the VM.

Do not store final release results on the VM including `/preproc/` dirs, the total

> size for all the recipes output, including `/preproc/` dirs is in the 4.5TB ballpark,

Author: good point! Just added a suggestion to use that or `remove_preproc_dir: true` in config

much too high for the VM storage capacity!

The steps to running the compare tool on the VM are the following:

- run date: log the run date here
- conda env: log the name of the conda environment you are using
- ESMValTool branch: log the name of the code branch you are using (e.g. `v2.8.x`)
- prerquisite - install `imagehash`: `pip install imagehash`
- reference run (v2.7.0): `export reference_dir=/work/bd0854/b382109/v270` (contains `preproc/` dirs too, 122 recipes)
- current run (v2.8.0): `export current_dir=path_to_current_run`
- command to run: `nohup python ESMValTool/esmvaltool/utils/testing/regression/compare.py --reference $reference_dir --current $current_dir > compare_v280_output.txt`

Some of the recipes will appear as having identical output to the one from previous release. However, others

> will need human inspection; here are some guidelines on how to perform the human inspection:

Reviewer: Should the human inspection be done by the release manager? I would try to ask recipe maintainers for this.

Author: err yes and no - that would delay the process, but it'd also accelerate it if peeps were on point - let's talk about it

Reviewer 2: I added some information to ask maintainers and the development community to support the RM

  as different - but Mark I eyeball inspection will show they are identical
- other plots will differ due to changes in plot settings (different colours, axes etc) due to updated settings from the
  diagnostic developers: if they look similar enough, then it's fine
- report (and subsequently open issues) if you notice major differences in plots; most times a simple comment on the
  release issue, whereby you tag the diagnostic developers leads to them having a look at the plots and OK-ing them; if that's
  not the case, then open a separate issue

Appendix
--------

Here you can find a list of useful files and directories:

- Formatted list of current recipes (as of v2.7.0) to be used with Markdown entries (on DKRZ/Lvante) at `/home/b/b382109/Tool_Release_270_Scripts/all_recipes.md` or in the draft Pull Request https://github.com/ESMValGroup/ESMValTool/pull/2883
- last release (v2.7.0) submit scripts on DKRZ/Levante `/home/b/b382109/submit`
- Miniconda3 installer file on DKRZ/Levante `/home/b/b382109/Miniconda3-py39_4.12.0-Linux-x86_64.sh` (remember to immediately update conda after using it, it is fairly old, from May 2022)

> - list of Autoassess reference files and masks on DKRZ/Levante `/home/b/b382109/autoassess_files`

Reviewer: This information seems prone to going out of date very quickly
- Our source code should be in a public GitHub repo
- Mamba installer can be downloaded from https://github.com/conda-forge/miniforge#mambaforge
- Data for running recipes should be publicly available. If that fails, it should at the very least be in a group workspace instead of someones home directory, but actually I would consider such recipes broken according to our broken recipe policy.

We should document various utilities' versions so that the work can be reproduced in case there
is an issue, or release work needs to be picked up mid-release by another release manager:

- documenting `conda`/`mamba` versions:

.. code-block:: bash

  mamba --version

- documenting `git` branch and its state:

.. code-block:: bash

  git status

Furthermore, the runtime environment needs to be documented: make a copy of the environment file,
and attach it in the release testing issue; to record the environment in a yaml file use e.g.

.. code-block:: bash

  conda env export > ToolEnv_v2xx_Test.txt

Modifications to configuration files need to be documented as well.
To test recipes, it is recommended to only use the default options and DKRZ data directories, simply by uncommenting
the DKRZ-Levante block of a newly generated ``config-user.yml`` file.

Submit run scripts - test recipe runs
-------------------------------------

We are now ready to start running all the available recipes, to compare output against previous release. Running is currently done
via batch scripts submitted to a schedulers (SLURM). Generate the submission scripts using the ``generate.py`` `Python script

> <https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/utils/batch-jobs/generate.py>`_. 

Reviewer: It would be nicer to link to the documentation describing this script instead of to the actual code

(note the `compute` partition which is used for such heavy runs):

.. code-block:: bash

  #SBATCH --partition=compute
  #SBATCH --time=08:00:00
  #SBATCH --constraint=512G

.. note::

  On DKRZ/Levante, a user can't have more than 20 SLURM jobs running at a time.
  As soon as a job is finished, the next one should start. More information on the job handling at DKRZ `here
  <https://docs.dkrz.de/doc/levante/running-jobs/partitions-and-limits.html#levante-partitions-and-limits>`_.

Analyse the results
-------------------

Once all jobs are completed, assemble some statistics so that issues with certain recipes
can be followed-up, and document this information in the release issue, such as:

- number of successfully run recipes
- number of failed recipes with preprocessor errors (can they be fixed? Can the fixes be included in the release?)
- number of failed recipes with diagnostic errors (can they be fixed? Can the fixes be included in the release?)
- number of recipes that are missing data
- number of recipes that have various other issues (and document them)

To parse the output of all these runs, use the ``parse_recipes_output.py`` `utility Python script

> <https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/utils/batch-jobs/parse_recipes_output.py>`_.

Reviewer: It would be nicer to link to the documentation describing this script instead of to the actual code

It is recommended to run the recipes with `log_level: info` in your config file to enable the parsing script to run fast.

Share the results with the community
------------------------------------

Create the debug.html and index.html overview webpages by running the utiliy script

> ``python esmvaltool/utils/testing/regression/summarize.py ~/esmvaltool_output/`` in the directory containing the recipe runs.

Reviewer: Please link to the documentation of the script instead of describing the script here

Reviewer 2: Done in 47257fdcb. I had to add move the docs on this script from the cylc section to a separate place.

These two files, together with the recipe output, need to be copied to the disk of a virtual machine (VM)
used to display recipe output in `webpages
<https://esmvaltool.dkrz.de/shared/esmvaltool/>`_.
Do not store final release results on the VM including `/preproc/` dirs, the total
size for all the recipes output, including `/preproc/` dirs is in the 4.5TB ballpark,
much too high for the VM storage capacity! Therefore, we would recommend using the option

> to remove preprocessing directories upon recipe running successfully `--remove-preproc-dir=True`

Reviewer: ```suggestion
to remove preprocessing directories upon recipe running successfully ``--remove-preproc-dir=True``
```
It is usual to use monospace to describe commands

> at runtime, or set `remove_preproc_dir: true` in the configuration file.

Reviewer: ```suggestion
at runtime, or set ``remove_preproc_dir: true`` in the configuration file.
```

Login and access to the DKRZ esmvaltool VM - results from recipe runs

> are stored on the VM; login with:

Reviewer: ```suggestion
are stored on the VM; log in to the Levante head node and then continue to the VM with:
```

.. code-block:: bash

  ssh user@esmvaltool.dkrz.de

where `user` is your DKRZ/Levante user name.
Then create a new subdirectory in ``/shared/esmvaltool/`` that will contain recipe output.
This should be named like the ESMValCore version used for the testing, e.g. ``v2.x.xrcx`` or  ``v2.x.x``.
Recipe output can be copied by doing from the VM:

.. code-block:: bash

  nohup cp -r /path_to_testing/esmvaltool_output/* /shared/esmvaltool/v2.x.x/
  
By copying the debug.html and index.html files into /shared/esmvaltool/v2.x.x/, the output
becomes available online, see for `example
<https://esmvaltool.dkrz.de/shared/esmvaltool/v2.7.0>`_.
Before copying the recipe output to the VM, you may want to clean up your directory containing

> the results and only keep the last run for each recipe.

Reviewer: ```suggestion
the results by removing any large `preproc` directories of failed runs and only keeping the last run for each recipe.
```

This will help generating a clearer overview webpage.
Note that the ``summarize.py`` script needs to be rerun if recipe runs were added or deleted
from your testing directory.

Link the overview webpage to the release issue.
This makes it much easier to ask for feedback from recipe developers and analyse failures.

Results produced with the final ESMValCore release candidate should be put in a VM directory
named after the version number, e.g. ``v2.x.x``. 
Once the release process is over, test results produced with previous release candidates can be deleted to save space on the VM.

.. note::

  If you wrote recipe runs output to Levante's `/scratch` partition, be aware that
  the data will be removed after two weeks, so you will have to quickly move the 

>   output data to the VM, using the `nohup` command above.

Reviewer: ```suggestion
  output data to the VM, using the ``nohup`` command above.
```

Running the comparison
----------------------

To compare the newly produced output from running all recipes, follow these steps below.

Access to the DKRZ esmvaltool VM, then install miniconda on the VM, and
if you have a Miniconda installer already downloaded in your Levante $HOME

.. code-block:: bash

  scp Miniconda3-py39_4.12.0-Linux-x86_64.sh user@esmvaltool.dkrz.de:/mnt/esmvaltool_disk2/work/<username>

.. warning::

  conda environments should not be created in the home directory because it is on a very small disk,
  but rather in a directory with your username under `/mnt/esmvaltool_disk2/work/<username>`

Next, we need to set up the input files

> The `/work` partition is visible by the VM so you can run the compare tool straight on the VM.

Reviewer: ```suggestion
The ``/work`` partition is visible from the VM so you can run the compare tool straight on the VM.
```

The steps to running the compare tool on the VM are the following:

- run date: log the run date here
- conda env: log the name of the conda environment you are using
- ESMValTool branch: log the name of the code branch you are using (e.g. `v2.8.x`)
- prerequisite - install `imagehash`: `pip install imagehash`
- reference run (v2.7.0; previous stable release): `export reference_dir=/work/bd0854/b382109/v270` (contains `preproc/` dirs too, 122 recipes)
- current run (v2.8.0): `export current_dir=path_to_current_run`
- command to run: 

> . code-block:: python

Reviewer: ```suggestion
.. code-block:: python
```

  nohup python ESMValTool/esmvaltool/utils/testing/regression/compare.py --reference $reference_dir --current $current_dir > compare_v280_output.txt

Copy the comparison txt file to the release issue.
Some of the recipes will appear as having identical output to the one from previous release. However, others
will need human inspection. Ask the recipe maintainers (`@ESMValGroup/esmvaltool-recipe-maintainers`_) and ESMValTool Development Team (`@ESMValGroup/esmvaltool-developmentteam`_) to provide assistance in checking the results.
Here are some guidelines on how to perform the human inspection:

- look at plots from current run vs previous release run: most of them will be identical, but if Matplotlib

>   has changed some plotting feature, images will have slightly different metadata so the comparison script will report them

Reviewer: It would be good to note here that the comparison is actually smart enough to overlook small differences, or even better link the documentation of the comparison script

>   has changed some plotting feature, images will have slightly different metadata so the comparison script will report them

Reviewer 2: Agreed. Done in 62e7f7e5e

  as different - but Mark I eyeball inspection will show they are identical
- other plots will differ due to changes in plot settings (different colours, axes etc) due to updated settings from the
  diagnostic developers: if they look similar enough, then it's fine

> - report (and subsequently open issues) if you notice major differences in plots; most times a simple comment on the

Reviewer 2: Yes, done in ab9af6d33bc51c56b4d844255ba85cb41912766d

  release issue, whereby you tag the diagnostic developers leads to them having a look at the plots and OK-ing them; if that's
  not the case, then open a separate issue

Appendix
--------

Here you can find a list of utility scripts used to run recipes and analyse the results:

- :ref:`Python scripts<utils_batch_jobs>` that create slurm submission scripts and parse slurm log files.
- :ref:`Python script<compare_recipe_runs>` that compares one or more recipe runs to known good previous run(s).

> - `Python summarize script <https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/utils/testing/regression/summarize.py>`_ that create the index.html overview page.

Reviewer: Link to the documentation of the script instead of to the code
